File: labs/J90_TemplateMethod/python/complete/bill.py
class Bill():
    id:str
    def __init__(self,id:str):
        self.id = id

    def _report(self,step):
        print('[bill:' + self.id + ' step:' + step + ']')

    def preProposal(self):
        pass

    def proposal(self):
        self._report("Proposal")

    def postProposal(self):
        pass

    def preCommittee(self):
        pass

    def committee(self):
        self._report("Committee Debate")
        self._report("Committee Vote")

    def postCommittee(self):
        pass

    def preFloor(self):
        pass

    def floor(self):
        self._report("Floor Debate")
        self._report("Floor Vote")

    def postFloor(self):
        pass

    def preSignoff(self):
        pass

    def signoff(self):
        self._report("Executive Sign-off")

    def postSignoff(self):
        pass

    def execute_workflow(self):
        self.preProposal()
        self.proposal()
        self.postProposal()

        self.preCommittee()
        self.committee()
        self.postCommittee()

        self.preFloor()
        self.floor()
        self.postFloor()

        self.preSignoff()
        self.signoff()
        self.postSignoff()

File: labs/J90_TemplateMethod/java/complete/app/src/main/java/TemplateMethodComplete/Bill.java
package TemplateMethodComplete;


public class Bill {
  private String id="";

  public Bill(String id){
    this.id=id;
  }

  public String getID(){
    return this.id;
  }

  protected void reportProgress(String step){
    System.out.print("[");
    System.out.print("Bill: " + this.id);
    System.out.print(", Step: " + step);
    System.out.println("]");
  }

  public void processWorkflow(){
		preProposal();
		proposal();
		postProposal();

		preCommittee();
		committee();
		postCommittee();

		preFloor();
		floor();
		postFloor();

		preSignoff();
		signoff();
		postSignoff();
  }

  protected void preProposal(){}
  protected void postProposal(){}
  protected void proposal(){
    this.reportProgress("Proposal");
  }


  protected void preCommittee(){}
  protected void postCommittee(){}
  protected void committee(){
    this.reportProgress("Committee debate");
    this.reportProgress("Committee vote");
  }


  protected void preFloor(){}
  protected void postFloor(){}
  protected void floor(){
    this.reportProgress("Floor debate");
    this.reportProgress("Floor vote");
  }


  protected void preSignoff(){}
  protected void postSignoff(){}
  protected void signoff(){
    this.reportProgress("Executive Signoff");
  }


}
